crates/forge_repo/src/proto/codebase_repository.rs

```diff
@@ -0,0 +1,251 @@
+            .context("Failed to parse workspace ID from server response")
+    }
+
+    async fn upload_files(&self, upload: &forge_domain::FileUpload) -> Result<UploadStats> {
+        let files: Vec<File> = upload
+            .data
+            .iter()
+            .map(|file_read| File {
+                path: file_read.path.clone(),
+                content: file_read.content.clone(),
+            })
+            .collect();
+
+        let request = UploadFilesRequest {
+            user_id: Some(UserId { id: upload.user_id.to_string() }),
+            workspace_id: Some(proto_generated::WorkspaceId {
+                id: upload.workspace_id.to_string(),
+            }),
+            content: Some(FileUploadContent { files, git: None }),
+        };
+
+        let mut client = self.client.clone();
+        let response = client.upload_files(request).await?;
+
+        let result = response
+            .into_inner()
+            .result
+            .context("Server did not return upload result in UploadFiles response")?;
+
+        Ok(UploadStats::new(result.nodes.len(), result.relations.len()))
+    }
+
+    /// Search for code using semantic search
+    async fn search(
+        &self,
+        search_query: &forge_domain::CodeSearchQuery<'_>,
+    ) -> Result<Vec<CodeSearchResult>> {
+        let request = tonic::Request::new(SearchRequest {
+            user_id: Some(UserId { id: search_query.user_id.to_string() }),
+            workspace_id: Some(proto_generated::WorkspaceId {
+                id: search_query.workspace_id.to_string(),
+            }),
+            query: Some(Query {
+                prompt: Some(search_query.data.query.to_string()),
+                limit: Some(search_query.data.limit as u32),
+                top_k: search_query.data.top_k,
+                ..Default::default()
+            }),
+        });
+
+        let mut client = self.client.clone();
+        let response = client.search(request).await?;
+
+        let result = response.into_inner().result.unwrap_or_default();
+
+        // Convert QueryItems to CodeSearchResults
+        let results = result
+            .data
+            .into_iter()
+            .filter_map(|query_item| {
+                let node = query_item.node?;
+                let node_data = node.data?;
+                let node_id = node.node_id.map(|n| n.id).unwrap_or_default();
+                let similarity = query_item.distance.unwrap_or(0.0);
+
+                // Convert proto node to domain CodeSearchResult based on type
+                let result = match node_data.kind? {
+                    node_data::Kind::FileChunk(chunk) => CodeSearchResult::FileChunk {
+                        node_id,
+                        file_path: chunk.path,
+                        content: chunk.content,
+                        start_line: chunk.start_line,
+                        end_line: chunk.end_line,
+                        similarity,
```

Comment:
similarity shouldn't be stored inside the node_data. Similarity should be stored outside.